arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedDeltaEncodedValuesReader.java

```diff
@@ -0,0 +1,276 @@
+    // First value
+    if (valuesRead == 0) {
+      outputWriter.write(vec, ((long) (currentRowId + valuesRead) * typeWidth), firstValue);
+      lastValueRead = firstValue;
+      currentRowId++;
+      remaining--;
+    }
+
+    while (remaining > 0) {
+      int loadedRows;
+      try {
+        loadedRows = loadMiniBlockToOutput(remaining, vec, currentRowId, typeWidth, outputWriter);
+      } catch (IOException e) {
+        throw new ParquetDecodingException("Error reading mini block.", e);
+      }
+      currentRowId += loadedRows;
+      remaining -= loadedRows;
+    }
+    valuesRead = total - remaining;
+  }
+
+  /**
+   * Read from a mini block. Read at most 'remaining' values into output.
+   *
+   * @return the number of values read into output
+   */
+  private int loadMiniBlockToOutput(
+      int remaining, FieldVector vec, int rowId, int typeWidth, IntegerOutputWriter outputWriter)
+      throws IOException {
+
+    // new block; read the block header
+    if (remainingInBlock == 0) {
+      readBlockHeader();
+    }
+
+    // new miniblock, unpack the miniblock
+    if (remainingInMiniBlock == 0) {
+      unpackMiniBlock();
+    }
+
+    // read values from miniblock
+    int valuesReadInMiniBlock = 0;
+    for (int i = miniBlockSizeInValues - remainingInMiniBlock;
+        i < miniBlockSizeInValues && valuesReadInMiniBlock < remaining;
+        i++) {
+      // calculate values from deltas unpacked for current block
+      long outValue = lastValueRead + minDeltaInCurrentBlock + unpackedValuesBuffer[i];
+      lastValueRead = outValue;
+      outputWriter.write(vec, ((long) (rowId + valuesReadInMiniBlock) * typeWidth), outValue);
+      remainingInBlock--;
+      remainingInMiniBlock--;
+      valuesReadInMiniBlock++;
+    }
+
+    return valuesReadInMiniBlock;
+  }
+
+  private void readBlockHeader() {
+    try {
+      minDeltaInCurrentBlock = BytesUtils.readZigZagVarLong(inputStream);
+    } catch (IOException e) {
+      throw new ParquetDecodingException("Can not read min delta in current block", e);
+    }
+    readBitWidthsForMiniBlocks();
+    remainingInBlock = blockSizeInValues;
+    currentMiniBlock = 0;
+    remainingInMiniBlock = 0;
+  }
+
+  /**
+   * mini block has a size of 8*n, unpack 32 value each time
+   *
+   * <p>see org.apache.parquet.column.values.delta.DeltaBinaryPackingValuesReader#unpackMiniBlock
+   */
+  private void unpackMiniBlock() throws IOException {
+    Arrays.fill(this.unpackedValuesBuffer, 0);
+    BytePackerForLong packer =
+        Packer.LITTLE_ENDIAN.newBytePackerForLong(bitWidths[currentMiniBlock]);
+    for (int j = 0; j < miniBlockSizeInValues; j += 8) {
+      ByteBuffer buffer = inputStream.slice(packer.getBitWidth());
+      if (buffer.hasArray()) {
+        packer.unpack8Values(
+            buffer.array(), buffer.arrayOffset() + buffer.position(), unpackedValuesBuffer, j);
+      } else {
+        packer.unpack8Values(buffer, buffer.position(), unpackedValuesBuffer, j);
```

Comment:
actually sorry that I missed this, but I was testing here and do we need the actual `buffer.hasArray()` condition? Unpacking the mini block into the bugger should always work no?

Reply:
Yeah I just looked in the upstream Spark changes, looks like it's mainly driven by performance; unpacking into the underlying array directly when possible (heap byte buffers?) is more performant than going through the regular buffer interface.